File: java/advanced/InlineAdaptiveBannerExample/app/src/main/java/com/google/android/gms/example/inlineadaptivebannerexample/MainActivity.java
package com.google.android.gms.example.inlineadaptivebannerexample;

import android.os.Build.VERSION;
import android.os.Build.VERSION_CODES;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.WindowMetrics;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.LoadAdError;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * A simple activity showing the use of {@link AdView} ads in a {@link RecyclerView} widget.
 *
 * <p>The {@link RecyclerView} widget is a more advanced and flexible version of ListView. This
 * widget helps simplify the display and handling of large data sets by allowing the layout manager
 * to determine when to reuse (recycle) item views that are no longer visible to the user. Recycling
 * views improves performance by avoiding the creation of unnecessary views or performing expensive
 * findViewByID() lookups.
 */
public class MainActivity extends AppCompatActivity {

  private static final String AD_UNIT_ID = "ca-app-pub-3940256099942544/1039341195";

  // A banner ad is placed in every 8th position in the RecyclerView.
  public static final int ITEMS_PER_AD = 8;

  // List of banner ads and MenuItems that populate the RecyclerView.
  private final List<Object> recyclerViewItems = new ArrayList<>();

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    // The RecyclerView that holds and displays banner ads and menu items.
    RecyclerView recyclerView = findViewById(R.id.recycler_view);

    // Use this setting to improve performance if you know that changes
    // in content do not change the layout size of the RecyclerView.
    recyclerView.setHasFixedSize(true);

    // Specify a linear layout manager.
    RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
    recyclerView.setLayoutManager(layoutManager);

    // Update the RecyclerView item's list with menu items and banner ads.
    addMenuItemsFromJson();
    addBannerAds();
    loadBannerAds();

    // Specify an adapter.
    RecyclerView.Adapter<RecyclerView.ViewHolder> adapter =
        new RecyclerViewAdapter(this, recyclerViewItems);
    recyclerView.setAdapter(adapter);
  }

  @Override
  protected void onResume() {
    for (Object item : recyclerViewItems) {
      if (item instanceof AdView) {
        AdView adView = (AdView) item;
        adView.resume();
      }
    }
    super.onResume();
  }

  @Override
  protected void onPause() {
    for (Object item : recyclerViewItems) {
      if (item instanceof AdView) {
        AdView adView = (AdView) item;
        adView.pause();
      }
    }
    super.onPause();
  }

  @Override
  protected void onDestroy() {
    for (Object item : recyclerViewItems) {
      if (item instanceof AdView) {
        AdView adView = (AdView) item;
        adView.destroy();
      }
    }
    super.onDestroy();
  }

  // Determine the screen width to use for the ad width.
  public int getAdWidth() {
    DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
    int adWidthPixels = displayMetrics.widthPixels;

    if (VERSION.SDK_INT >= VERSION_CODES.R) {
      WindowMetrics windowMetrics = this.getWindowManager().getCurrentWindowMetrics();
      adWidthPixels = windowMetrics.getBounds().width();
    }

    float density = displayMetrics.density;
    return (int) (adWidthPixels / density);
  }

  /** Adds banner ads to the items list. */
  private void addBannerAds() {
    // Loop through the items array and place a new banner ad in every ith position in
    // the items List.
    for (int i = 0; i <= recyclerViewItems.size(); i += ITEMS_PER_AD) {
      final AdView adView = new AdView(MainActivity.this);
      adView.setAdSize(AdSize.getCurrentOrientationInlineAdaptiveBannerAdSize(this, getAdWidth()));
      adView.setAdUnitId(AD_UNIT_ID);
      recyclerViewItems.add(i, adView);
    }
  }

  /** Sets up and loads the banner ads. */
  private void loadBannerAds() {
    // Load the first banner ad in the items list (subsequent ads will be loaded automatically
    // in sequence).
    loadBannerAd(0);
  }

  /** Loads the banner ads in the items list. */
  private void loadBannerAd(final int index) {

    if (index >= recyclerViewItems.size()) {
      return;
    }

    Object item = recyclerViewItems.get(index);
    if (!(item instanceof AdView)) {
      throw new ClassCastException(
          "Expected item at index " + index + " to be a banner ad" + " ad.");
    }

    final AdView adView = (AdView) item;

    // Set an AdListener on the AdView to wait for the previous banner ad
    // to finish loading before loading the next ad in the items list.
    adView.setAdListener(
        new AdListener() {
          @Override
          public void onAdLoaded() {
            super.onAdLoaded();
            // The previous banner ad loaded successfully, call this method again to
            // load the next ad in the items list.
            loadBannerAd(index + ITEMS_PER_AD);
          }

          @Override
          public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
            // The previous banner ad failed to load. Call this method again to load
            // the next ad in the items list.
            String error =
                String.format(
                    Locale.US,
                    "domain: %s, code: %d, message: %s",
                    loadAdError.getDomain(),
                    loadAdError.getCode(),
                    loadAdError.getMessage());
            Log.e(
                "MainActivity",
                "The previous banner ad failed to load with error: "
                    + error
                    + ". Attempting to"
                    + " load the next banner ad in the items list.");
            loadBannerAd(index + ITEMS_PER_AD);
          }
        });

    // Load the banner ad.
    adView.loadAd(new AdRequest.Builder().build());
  }

  /** Adds {@link MenuItem}'s from a JSON file. */
  private void addMenuItemsFromJson() {
    try {
      String jsonDataString = readJsonDataFromFile();
      JSONArray menuItemsJsonArray = new JSONArray(jsonDataString);

      for (int i = 0; i < menuItemsJsonArray.length(); ++i) {

        JSONObject menuItemObject = menuItemsJsonArray.getJSONObject(i);

        String menuItemName = menuItemObject.getString("name");
        String menuItemDescription = menuItemObject.getString("description");
        String menuItemPrice = menuItemObject.getString("price");
        String menuItemCategory = menuItemObject.getString("category");
        String menuItemImageName = menuItemObject.getString("photo");

        MenuItem menuItem =
            new MenuItem(
                menuItemName,
                menuItemDescription,
                menuItemPrice,
                menuItemCategory,
                menuItemImageName);
        recyclerViewItems.add(menuItem);
      }
    } catch (IOException | JSONException exception) {
      Log.e(MainActivity.class.getName(), "Unable to parse JSON file.", exception);
    }
  }

  /**
   * Reads the JSON file and converts the JSON data to a {@link String}.
   *
   * @return A {@link String} representation of the JSON data.
   * @throws IOException if unable to read the JSON file.
   */
  private String readJsonDataFromFile() throws IOException {

    InputStream inputStream = null;
    StringBuilder builder = new StringBuilder();

    try {
      String jsonDataString = null;
      inputStream = getResources().openRawResource(R.raw.menu_items_json);
      BufferedReader bufferedReader =
          new BufferedReader(new InputStreamReader(inputStream, "UTF-8"));
      while ((jsonDataString = bufferedReader.readLine()) != null) {
        builder.append(jsonDataString);
      }
    } finally {
      if (inputStream != null) {
        inputStream.close();
      }
    }

    return new String(builder);
  }
}

File: kotlin/advanced/InlineAdaptiveBannerExample/app/src/main/java/com/google/android/gms/example/inlineadaptivebannerexample/MainActivity.kt
package com.google.android.gms.example.inlineadaptivebannerexample

import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.WindowMetrics
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.RequestConfiguration
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONException

/** A simple activity showing the use of [AdView] ads in a [RecyclerView] widget. */
class MainActivity : AppCompatActivity() {
  // The RecyclerView that holds and displays banner ads and menu items.

  // List of banner ads and MenuItems that populate the RecyclerView.
  private val recyclerViewItems: MutableList<Any> = ArrayList()

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    // Set your test devices.
    MobileAds.setRequestConfiguration(
      RequestConfiguration.Builder().setTestDeviceIds(listOf(TEST_DEVICE_HASHED_ID)).build()
    )

    CoroutineScope(Dispatchers.IO).launch {
      // Initialize the Google Mobile Ads SDK on a background thread.
      MobileAds.initialize(this@MainActivity) {}
    }

    val recyclerView: RecyclerView? = findViewById(R.id.recycler_view)

    // The size of the RecyclerView depends on the height of the ad.
    recyclerView?.setHasFixedSize(false)

    // Specify a linear layout manager.
    val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
    recyclerView?.layoutManager = layoutManager

    // Update the RecyclerView item's list with menu items and banner ads.
    addMenuItemsFromJson()
    addBannerAds()
    loadBannerAds()

    // Specify an adapter.
    val adapter: RecyclerView.Adapter<ViewHolder> = RecyclerViewAdapter(this, recyclerViewItems)
    recyclerView?.adapter = adapter
  }

  // Determine the screen width to use for the ad width.
  // [START get_ad_width]
  private val adWidth: Int
    get() {
      val displayMetrics = resources.displayMetrics
      val adWidthPixels =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
          val windowMetrics: WindowMetrics = this.windowManager.currentWindowMetrics
          windowMetrics.bounds.width()
        } else {
          displayMetrics.widthPixels
        }
      val density = displayMetrics.density
      return (adWidthPixels / density).toInt()
    }

  // [END get_ad_width]

  override fun onResume() {
    for (item in recyclerViewItems) {
      if (item is AdView) {
        item.resume()
      }
    }
    super.onResume()
  }

  override fun onPause() {
    for (item in recyclerViewItems) {
      if (item is AdView) {
        item.pause()
      }
    }
    super.onPause()
  }

  override fun onDestroy() {
    for (item in recyclerViewItems) {
      if (item is AdView) {
        item.destroy()
      }
    }
    super.onDestroy()
  }

  /** Adds banner ads to the items list. */
  private fun addBannerAds() {
    // Loop through the items array and place a new banner ad in every ith position in
    // the items List.
    var i = 0
    while (i <= recyclerViewItems.size) {
      // [START create_banner_ad_view]
      val adView = AdView(this@MainActivity)
      adView.setAdSize(AdSize.getCurrentOrientationInlineAdaptiveBannerAdSize(this, adWidth))
      // [END create_banner_ad_view]
      adView.adUnitId = AD_UNIT_ID
      recyclerViewItems.add(i, adView)
      i += ITEMS_PER_AD
    }
  }

  /** Sets up and loads the banner ads. */
  private fun loadBannerAds() {
    // Load the first banner ad in the items list (subsequent ads will be loaded automatically
    // in sequence).
    loadBannerAd(0)
  }

  /** Loads the banner ads in the items list. */
  private fun loadBannerAd(index: Int) {
    if (index >= recyclerViewItems.size) {
      return
    }
    val item =
      recyclerViewItems[index] as? AdView
        ?: throw ClassCastException("Expected item at index $index to be a banner ad ad.")

    // Set an AdListener on the AdView to wait for the previous banner ad
    // to finish loading before loading the next ad in the items list.
    item.adListener =
      object : AdListener() {
        override fun onAdLoaded() {
          super.onAdLoaded()
          // The previous banner ad loaded successfully, call this method again to
          // load the next ad in the items list.
          loadBannerAd(index + ITEMS_PER_AD)
        }

        override fun onAdFailedToLoad(loadAdError: LoadAdError) {
          // The previous banner ad failed to load. Call this method again to load
          // the next ad in the items list.
          val error =
            String.format(
              "domain: %s, code: %d, message: %s",
              loadAdError.domain,
              loadAdError.code,
              loadAdError.message,
            )
          Log.e(
            "MainActivity",
            "The previous banner ad failed to load with error: " +
              error +
              ". Attempting to" +
              " load the next banner ad in the items list.",
          )
          loadBannerAd(index + ITEMS_PER_AD)
        }
      }

    // Load the banner ad.
    item.loadAd(AdRequest.Builder().build())
  }

  /** Adds [MenuItem]'s from a JSON file. */
  private fun addMenuItemsFromJson() {
    try {
      val jsonDataString = readJsonDataFromFile()
      val menuItemsJsonArray = JSONArray(jsonDataString)
      for (i in 0 until menuItemsJsonArray.length()) {
        val menuItemObject = menuItemsJsonArray.getJSONObject(i)
        val menuItemName = menuItemObject.getString("name")
        val menuItemDescription = menuItemObject.getString("description")
        val menuItemPrice = menuItemObject.getString("price")
        val menuItemCategory = menuItemObject.getString("category")
        val menuItemImageName = menuItemObject.getString("photo")
        val menuItem =
          MenuItem(
            menuItemName,
            menuItemDescription,
            menuItemPrice,
            menuItemCategory,
            menuItemImageName,
          )
        recyclerViewItems.add(menuItem)
      }
    } catch (exception: IOException) {
      Log.e(MainActivity::class.java.name, "Unable to parse JSON file.", exception)
    } catch (exception: JSONException) {
      Log.e(MainActivity::class.java.name, "Unable to parse JSON file.", exception)
    }
  }

  /**
   * Reads the JSON file and converts the JSON data to a [String].
   *
   * @return A [String] representation of the JSON data.
   * @throws IOException if unable to read the JSON file.
   */
  @Throws(IOException::class)
  private fun readJsonDataFromFile(): String {
    var inputStream: InputStream? = null
    val builder = StringBuilder()
    try {
      var jsonDataString: String? = null
      inputStream = resources.openRawResource(R.raw.menu_items_json)
      val bufferedReader = BufferedReader(InputStreamReader(inputStream, "UTF-8"))
      while (bufferedReader.readLine().also { jsonDataString = it } != null) {
        builder.append(jsonDataString)
      }
    } finally {
      inputStream?.close()
    }
    return builder.toString()
  }

  companion object {
    private const val AD_UNIT_ID = "ca-app-pub-3940256099942544/1039341195"

    // A banner ad is placed in every 8th position in the RecyclerView.
    const val ITEMS_PER_AD = 8

    // Check your logcat output for the test device hashed ID e.g.
    // "Use RequestConfiguration.Builder().setTestDeviceIds(Arrays.asList("ABCDEF012345"))
    // to get test ads on this device" or
    // "Use new ConsentDebugSettings.Builder().addTestDeviceHashedId("ABCDEF012345") to set this as
    // a debug device".
    const val TEST_DEVICE_HASHED_ID = "ABCDEF012345"
  }
}
